### PR TITLE
Added an option to deploy all files in the test source location

### PR DESCRIPTION
**Pull request by cltshivash**
Added an option to deploy the test source dependencies and satellite dlls by default (DeployTestSourceDependencies) in addition to the content of the test source directory (bin/debug of the test dll). This option will be on by default for compat reasons.

It can be turned off in which case deployment would only deploy the files from the test source location.

Added UTs and E2E tests for the change

### CODE REVIEW

File: test/UnitTests/PlatformServices.Desktop.Unit.Tests/Utilities/DeploymentUtilityTests.cs [hunk 1]
```diff
@@ -364,6 +366,43 @@ public void DeployShouldDeployContentsOfADirectoryIfSpecified()
                 Times.Once);
         }
 
+        [TestMethod]
+        public void FileUtilityTests()
+        {
+        }
```
Comment (cltshivash) [inline test/UnitTests/PlatformServices.Desktop.Unit.Tests/Utilities/DeploymentUtilityTests.cs]: will remove this

File: test/UnitTests/PlatformServices.Desktop.Unit.Tests/Utilities/DeploymentUtilityTests.cs [hunk 2]
```diff
@@ -364,6 +366,43 @@ public void DeployShouldDeployContentsOfADirectoryIfSpecified()
...
+
+        [TestMethod]
+        public void DeployTestRunSourceOptionShouldDeployAllFilesAndDirectoriesPresentInSourceDirectory()
```
Comment (cltshivash) [inline test/UnitTests/PlatformServices.Desktop.Unit.Tests/Utilities/DeploymentUtilityTests.cs]: >DeployTestRunSourceOptionShouldDeployAllFilesAndDirectoriesPresentInSourceDirectory [](start = 20, length = 83)

same here.. i have added an E2E test  #Resolved

File: src/Adapter/PlatformServices.Desktop/Services/MSTestAdapterSettings.cs [hunk 3]
```diff
@@ -97,6 +34,11 @@ public MSTestAdapterSettings()
         /// </summary>
         public bool DeleteDeploymentDirectoryAfterTestRunIsComplete { get; private set; }
 
+        /// <summary>
+        /// Gets a value indicating whether the test source references are to deployed
```
Comment (jayaranigarg) [inline src/Adapter/PlatformServices.Desktop/Services/MSTestAdapterSettings.cs:38]: Nit: are to **be** deployed.

File: src/Adapter/PlatformServices.Desktop/Services/MSTestAdapterSettings.cs [hunk 4]
```diff
@@ -116,6 +57,7 @@ public static MSTestAdapterSettings ToSettings(XmlReader reader)
             //
             // <MSTestV2>
             //     <DeploymentEnabled>true</DeploymentEnabled>
+            //     <DeployTestSourceDependencies>true</DeployTestSourceDependencies>
```
Comment (jayaranigarg) [inline src/Adapter/PlatformServices.Desktop/Services/MSTestAdapterSettings.cs:60]: Nice 👍

File: src/Adapter/PlatformServices.Desktop/PlatformServices.Desktop.csproj [hunk 5]
```diff
@@ -77,7 +79,7 @@
     <Compile Include="Services\DesktopReflectionOperations.cs" />
     <Compile Include="Services\DesktopThreadOperations.cs" />
     <Compile Include="Services\DesktopFileOperations.cs" />
-    <Compile Include="Services\DesktopSettingsProvider.cs" />
+    <Compile Include="Services\MSTestSettingsProvider.cs" />
```
Comment (jayaranigarg) [inline src/Adapter/PlatformServices.Desktop/PlatformServices.Desktop.csproj:81]: We have different SettingsProvider for each platform service and their naming is kept in a fashion like Platform+SettingsProvider.
So for desktop PlatformService, it was DesktopSettingsProvider, netcore it is ns10SettingsProvider.
Kindly consider changing it back to DesktopSettingsProvider to keep the nomenclature consistent across. #Resolved

File: src/Adapter/PlatformServices.Desktop/Utilities/DeploymentItemUtility.cs [hunk 6]
```diff
@@ -91,14 +92,14 @@ internal bool IsValidDeploymentItem(string sourcePath, string relativeOutputDire
                 return false;
             }
 
-            if (sourcePath.IndexOfAny(System.IO.Path.GetInvalidPathChars()) != -1 ||
-                relativeOutputDirectory.IndexOfAny(System.IO.Path.GetInvalidPathChars()) != -1)
+            if (sourcePath.IndexOfAny(Path.GetInvalidPathChars()) != -1 ||
```
Comment (jayaranigarg) [inline src/Adapter/PlatformServices.Desktop/Utilities/DeploymentItemUtility.cs:95]: 👍

File: src/Adapter/PlatformServices.Desktop/Utilities/DeploymentResult.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+namespace Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Microsoft.VisualStudio.TestPlatform.MSTestAdapter.PlatformServices.Deployment;
+
+    internal class DeploymentResult
```
Comment (jayaranigarg) [inline src/Adapter/PlatformServices.Desktop/Utilities/DeploymentResult.cs]: Adding comments summarizing purpose of this class would be good.

**Comment by singhsarab** [comment]
>                 RunDirectories = runDirectories;

This updation is not clear. Do we still need this?

---
Refers to: src/Adapter/PlatformServices.Desktop/Services/DesktopTestDeployment.cs:158 in a43a940. [](commit_id = a43a9404f586ab44db2b8c22c59a9986b7f0f41c, deletion_comment = False)

File: src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs [hunk 8]
```diff
@@ -156,27 +144,37 @@ private bool Deploy(string source, IRunContext runContext, ITestExecutionRecorde
...
         [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Requirement is to handle all kinds of user exceptions and message appropriately.")]
-        private void Deploy(IList<DeploymentItem> deploymentItems, string testSource, string deploymentDirectory, out IEnumerable<string> deploymentWarnings)
+        private IEnumerable<string> Deploy(IList<DeploymentItem> deploymentItems, string testSource, string deploymentDirectory, string resultsDirectory)
         {
-            Debug.Assert(!string.IsNullOrEmpty(deploymentDirectory), "Deployment directory is null/empty");
-            Debug.Assert(this.fileUtility.DoesDirectoryExist(deploymentDirectory), "Deployment directory " + deploymentDirectory + " does not exist");
-            Debug.Assert(!string.IsNullOrEmpty(testSource), "TestSource directory is null/empty");
-            Debug.Assert(this.fileUtility.DoesFileExist(testSource), "TestSource " + testSource + " does not exist.");
+            Validate.IsFalse(string.IsNullOrWhiteSpace(deploymentDirectory), "Deployment directory is null or empty");
```
Comment (singhsarab) [inline src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs:152]: >Validate [](start = 12, length = 8)

Just thinking if we could reuse ValidateArg here.
Should we add tests for these validation cases as well ?

File: src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs [hunk 9]
```diff
@@ -156,27 +144,37 @@ private bool Deploy(string source, IRunContext runContext, ITestExecutionRecorde
...
+            Validate.IsFalse(string.IsNullOrWhiteSpace(testSource), "TestSource directory is null/empty");
+            Validate.IsTrue(this.fileUtility.DoesFileExist(testSource), $"TestSource {testSource} does not exist.");
 
             testSource = Path.GetFullPath(testSource);
-
             var warnings = new List<string>();
 
-            // Get the referenced assemblies.
-            this.ProcessNewStorage(testSource, deploymentItems, warnings);
+            if (MSTestSettingsProvider.Settings.DeployTestSourceDependencies)
```
Comment (singhsarab) [inline src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs:160]: Just confirming, We are deploying only the dependencies if this is enabled else we deploy items in the test source folder ?

File: src/Adapter/PlatformServices.Desktop/Services/DesktopTestDeployment.cs [hunk 10]
```diff
@@ -153,7 +151,7 @@ public bool Deploy(IEnumerable<TestCase> tests, IRunContext runContext, IFramewo
                 foreach (var group in testsBySource)
                 {
                     // do the deployment
-                    isDeploymentDone = this.deploymentUtility.Deploy(@group.Tests, @group.Source, runContext, frameworkHandle, ref runDirectories) || isDeploymentDone;
+                    this.deploymentUtility.Deploy(@group.Tests, @group.Source, runContext, frameworkHandle, RunDirectories);
```
Comment (jayaranigarg) [inline src/Adapter/PlatformServices.Desktop/Services/DesktopTestDeployment.cs:154]: Do we need 'runDirectories' after this change?

File: src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs [hunk 11]
```diff
@@ -156,27 +144,37 @@ private bool Deploy(string source, IRunContext runContext, ITestExecutionRecorde
...
-            var satelliteItems = this.GetSatellites(deploymentItems, testSource, warnings);
-            foreach (var satelliteItem in satelliteItems)
+                // Get the satellite assemblies
+                var satelliteItems = this.GetSatellites(deploymentItems, testSource, warnings);
+                foreach (var satelliteItem in satelliteItems)
+                {
+                    this.deploymentItemUtility.AddDeploymentItem(deploymentItems, satelliteItem);
+                }
+            }
+            else if (File.Exists(testSource))
```
Comment (jayaranigarg) [inline src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs]: We already have this check in beginning of this Deploy() function. Probably we can skip it here.

File: src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs [hunk 12]
```diff
@@ -121,32 +117,24 @@ private static void LogWarnings(ITestExecutionRecorder testExecutionRecorder, IE
...
                 foreach (var item in deploymentItems)
                 {
                     EqtTrace.Info("MSTestExecutor: SourcePath: - {0}", item.SourcePath);
                 }
             }
 
             // Do the deployment.
-            IEnumerable<string> warnings;
-
-            var runDirectories = testRunDirectories ?? this.CreateDeploymentDirectories(runContext);
```
Comment (jayaranigarg) [inline src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs:140]: 👍

**Comment by jayaranigarg** [comment]
Please put up an RFC in testfx-docs describing the usage of this 'DeployTestSourceDependencies' flag.

**Comment by jayaranigarg** [comment]
@cltshivash : Please complete this PR.

File: src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs (hunk 8 above)
Comment (cltshivash) [reply src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs:152]: It is a private function and hence not adding it.. ideally these arg validations are redundant as well. Will pick up later

File: src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs (hunk 9 above)
Comment (cltshivash) [reply src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs:160]: That's correct.

File: src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs (hunk 11 above)
Comment (cltshivash) [reply src/Adapter/PlatformServices.Desktop/Utilities/DeploymentUtility.cs]: Done.

**Pull request merged**